quic/s2n-quic-transport/src/ack/ack_ranges.rs

```diff
@@ -0,0 +1,241 @@
+
+impl Default for AckRanges {
+    fn default() -> Self {
+        Self::new(Settings::default().ack_ranges_limit as usize)
+    }
+}
+
+impl AckRanges {
+    pub fn new(limit: usize) -> Self {
+        let limit = NonZeroUsize::new(limit).expect("limit should be nonzero");
+        Self(IntervalSet::with_limit(limit))
+    }
+
+    /// Inserts a packet number; dropping smaller values if needed
+    #[inline]
+    pub fn insert_packet_number_range(&mut self, pn_range: PacketNumberRange) -> Result<(), ()> {
+        let interval = (
+            Bound::Included(pn_range.start()),
+            Bound::Included(pn_range.end()),
+        );
+        if self.0.insert(interval).is_ok() {
+            return Ok(());
+        }
+
+        // attempt to shed the lowest packet number ranges to make room for larger ones
+        match self.0.pop_min() {
+            Some(min) => {
+                if min < pn_range.start() {
+                    // TODO: add metrics for ack ranges being dropped
+                    self.0.insert(interval).expect(
+                        "min range was removed, so it should be possible to insert another range",
+                    );
+                    Ok(())
+                } else {
+                    // new value is smaller than min so inset it back in the front
+                    let _ = self.0.insert_front(min);
+                    Err(())
+                }
+            }
+            None => Err(()),
+        }
+    }
+
+    /// Inserts a packet number; dropping smaller values if needed
+    #[inline]
+    pub fn insert_packet_number(&mut self, packet_number: PacketNumber) -> Result<(), ()> {
+        self.insert_packet_number_range(PacketNumberRange::new(packet_number, packet_number))
+    }
+
+    /// Returns the overall range of the ack_ranges
+    #[inline]
+    pub fn spread(&self) -> usize {
+        match (self.min_value(), self.max_value()) {
+            (Some(min), Some(max)) => {
+                let min = PacketNumber::as_varint(min);
+                let max = PacketNumber::as_varint(max);
+                (max - min).try_into().unwrap_or(core::usize::MAX)
+            }
+            _ => 0,
+        }
+    }
+}
+
+type AckRangesIter<'a> = core::iter::Map<
+    core::iter::Rev<RangeInclusiveIter<'a, PacketNumber>>,
+    fn(RangeInclusive<PacketNumber>) -> RangeInclusive<VarInt>,
+>;
+
+impl<'a> ack::AckRanges for &'a AckRanges {
+    type Iter = AckRangesIter<'a>;
+
+    fn ack_ranges(&self) -> Self::Iter {
+        self.0.inclusive_ranges().rev().map(|range| {
+            let (start, end) = range.into_inner();
+            PacketNumber::as_varint(start)..=PacketNumber::as_varint(end)
+        })
+    }
+}
+
+impl Deref for AckRanges {
+    type Target = IntervalSet<PacketNumber>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl DerefMut for AckRanges {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
+    }
+}
+
+#[cfg(test)]
+pub mod tests {
```

Comment:
```suggestion
mod tests {
```

Reply:
Also I think a simple bolero test that inserts packet number ranges into the set would be useful to catch any panics.